tests/src/test/java/org/eclipse/hono/tests/registry/CredentialsApiTests.java

```diff
@@ -183,6 +184,65 @@ public void testGetCredentialsWithAutoProvisioning(final VertxTestContext ctx)
                 }));
     }
 
+    /**
+     * Verifies when no credentials are found, the properties related to auto-provisioning of gateways are enabled
+     * in the corresponding tenant's CA entry and the client context contains a serialized X.509 certificate then 
+     * a gateway is auto-provisioned. (i.e A gateway is registered and it's corresponding credentials are stored). 
+     *
+     * @param ctx The vert.x test context.
+     * @throws CertificateException if the self signed certificate cannot be created.
+     * @throws FileNotFoundException if the self signed certificate cannot be read.
+     */
+    @Timeout(value = 5, timeUnit = TimeUnit.SECONDS)
+    @Test
+    public void testGetCredentialsWithGatewayAutoProvisioning(final VertxTestContext ctx)
+            throws CertificateException, FileNotFoundException {
+
+        // GIVEN a tenant's trusted CA entry with auto-provisioning and auto-provision as gateway enabled
+        // and a client context that contains a client certificate
+        // while device has not been registered and no credentials are stored yet
+        final X509Certificate cert = createCertificate();
+        final var tenant = Tenants.createTenantForTrustAnchor(cert);
+        tenant.getTrustedCertificateAuthorities()
+                .get(0)
+                .setAutoProvisioningEnabled(true)
+                .setAutoProvisioningAsGatewayEnabled(true);
+        final JsonObject clientCtx = new JsonObject().put(CredentialsConstants.FIELD_CLIENT_CERT, cert.getEncoded());
+        final String authId = cert.getSubjectX500Principal().getName(X500Principal.RFC2253);
+
+        tenantId = getHelper().getRandomTenantId();
+        getHelper().registry
+                .addTenant(tenantId, tenant)
+                .compose(ok -> getClient()
+                        // WHEN getting credentials
+                        .get(tenantId, CredentialsConstants.SECRETS_TYPE_X509_CERT, authId, clientCtx, spanContext))
+                .onComplete(ctx.succeeding(result -> {
```

Comment:
how about changing this into a `compose` invocation doing all the assertions and returning the output of
```
getHelper().registry.getRegistrationInfo(tenantId, result.getDeviceId());
```
and then do the final assertions in an *onComplete* block?
FMPOV that would improve readability because it prevents the nesting of the async processing ...

Reply:
I have pushed the changes in accordance with the comments. Would you mind taking a look?